CommonMC/src/EventWindowMarkerProducer_module.cc

```diff
@@ -26,42 +29,114 @@ namespace mu2e {
       CLHEP::RandFlat _randflat;
+
+      double _initialPhaseShift;
   };
 
-  EventWindowMarkerProducer::EventWindowMarkerProducer(fhicl::ParameterSet const& pset):
-    art::EDProducer{pset},
-    _systemClockSpeed(pset.get<float>("SystemClockSpeed",40.0)), // MHz
-    _eventWindowConstantOffset(pset.get<float>("EventWindowConstantOffset",0.0)), // ns (is a calibration value to get data/sim agreement)
-    // a positive value means the event window is delayed relative to the StepPointMC timing
+  EventWindowMarkerProducer::EventWindowMarkerProducer(Parameters const& config):
+    art::EDProducer{config},
+    _systemClockSpeed( config().systemClockSpeed()), // MHz
+    _constantOffset( config().constantOffset()), // ns (is a calibration value to get data/sim agreement)
+    _spillType(static_cast<EventWindowMarker::SpillType>( config().spillType())),
+    _offSpillLength( config().offSpillLength()),
+    _onSpillBins( config().onSpillBins()),
+    _onSpillMaxLength( config().onSpillMaxLength()),
+    _recoFromMCTruth( config().recoFromMCTruth()),
+    _recoFromMCTruthErr( config().recoFromMCTruthErr()),
     _engine(createEngine( art::ServiceHandle<SeedService>()->getSeed())),
+    _randgauss( _engine ),
     _randflat( _engine )
   {
     produces<EventWindowMarker>();
+    produces<ProtonBunchTimeMC>();
+    produces<ProtonBunchTime>();
   }
 
   void EventWindowMarkerProducer::beginJob() {
+    // Proton bunch is 0 to 25 ns before event window marker
+    _initialPhaseShift = -1*_randflat.fire(1/_systemClockSpeed*1000);
   }
 
   void EventWindowMarkerProducer::beginRun(art::Run& run) {
   }
 
   void EventWindowMarkerProducer::produce(art::Event& event) {
       unique_ptr<EventWindowMarker> marker(new EventWindowMarker);
-      marker.get()->_timeOffset = _eventWindowConstantOffset + _randflat.fire(1/_systemClockSpeed*1000) - 1/_systemClockSpeed*1000/2.; // ns
+      unique_ptr<ProtonBunchTimeMC> pbtmc(new ProtonBunchTimeMC);
+      unique_ptr<ProtonBunchTime> pbt(new ProtonBunchTime);
+      marker.get()->_spillType = _spillType;
+
+      double period = 1/_systemClockSpeed*1000;
+
+      if (_spillType == EventWindowMarker::SpillType::offspill){
+        pbtmc.get()->pbtime_ = 0;
+        pbt.get()->pbtime_ = 0;
+        pbt.get()->pbterr_ = 0;
+        marker.get()->_eventLength = _offSpillLength*period;
+      }else{
+        // calculate which bin we are in from event number
+        int bin = event.id().event() % _onSpillBins;
+        // determine phase for this event
+        // each microbunch the proton bunch gets shifted back 5 ns from the event window
+        double shiftPer = -1*period/_onSpillBins;
+        double phaseShift = (_initialPhaseShift + bin*shiftPer);
+        if (phaseShift < -1*period)
+          phaseShift += period;
+        // when phase shift would be > than one clock period by next microbunch
+        // the event window is one clock tick shorter
+        if (phaseShift < -1*period - shiftPer)
```

Comment:
isn't it always the case that this logic will pick bin 4 for the 67 cycle event?  If so, it might be clearer to code it as such.

Reply:
I was just trying to not have any magic numbers, I can hard code that if we think its ok

Reply:
If this mimics what the actual hardware does then it's OK.  As follow-on to this PR we should add some validation checks that this behaves as expected.